File: DatabindingWithNewsApi/app/src/main/java/com/enpassio/databindingwithnewsapi/ui/NewsAdapter.java
 package com.enpassio.databindingwithnewsapi.ui;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.enpassio.databindingwithnewsapi.R;
import com.enpassio.databindingwithnewsapi.databinding.ItemBinding;
import com.enpassio.databindingwithnewsapi.model.Article;

import java.util.List;


public class NewsAdapter extends RecyclerView.Adapter<NewsAdapter.NewsViewHolder> {

    private List<Article> mArticleList;
    private final ArticleClickListener mListener;

    NewsAdapter(ArticleClickListener listener) {
        mListener = listener;
    }

    void setArticleList(List<Article> articleList) {
        mArticleList = articleList;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public NewsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemBinding binding = DataBindingUtil
                .inflate(LayoutInflater.from(parent.getContext()), R.layout.item_news_article,
                        parent, false);

        return new NewsViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull NewsViewHolder holder, int position) {
        holder.bind(mArticleList.get(position), mListener);
    }

    @Override
    public int getItemCount() {
        //If list is null, return 0, otherwise return the size of the list
        return mArticleList == null ? 0 : mArticleList.size();
    }

    class NewsViewHolder extends RecyclerView.ViewHolder {

        final ItemBinding binding;

        NewsViewHolder(ItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(Article currentArticle, ArticleClickListener clickListener){
            //For each item, corresponding product object is passed to the binding
            binding.setArticle(currentArticle);
            //Pass an item click listener to each item layout.
            binding.setArticleItemClick(clickListener);
            //This is to force bindings to execute right away
            binding.executePendingBindings();
        }
    }

    public interface ArticleClickListener {
        void onArticleClicked(Article chosenArticle);
    }
}

File: DatabindingWithNewsApiKotlin/app/src/main/java/com/enpassio/databindingwithnewsapikotlin/ui/NewsAdapter.kt
package com.enpassio.databindingwithnewsapikotlin.ui

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.enpassio.databindingwithnewsapikotlin.R
import com.enpassio.databindingwithnewsapikotlin.databinding.ItemBinding
import com.enpassio.databindingwithnewsapikotlin.model.Article


class NewsAdapter (private val mListener: ArticleClickListener) :
    RecyclerView.Adapter<NewsAdapter.NewsViewHolder>() {

    var articleList: List<Article>? = null
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsViewHolder = NewsViewHolder.from(parent)

    override fun onBindViewHolder(holder: NewsViewHolder, position: Int) = holder.bind(articleList?.get(position), mListener)

    override fun getItemCount(): Int {
        //If list is null, return 0, otherwise return the size of the list
        return articleList?.size ?: 0
    }

    class NewsViewHolder(val binding: ItemBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(currentArticle: Article?, clickListener: ArticleClickListener){
            binding.article = currentArticle
            binding.articleItemClick = clickListener
            binding.executePendingBindings()
        }

        companion object {
            fun from(parent: ViewGroup): NewsViewHolder {
                val binding = DataBindingUtil.inflate<ItemBinding>(LayoutInflater.from(parent.context),
                    R.layout.item_news_article, parent, false)
                return NewsViewHolder(binding)
            }
        }
    }

    interface ArticleClickListener {
        fun onArticleClicked(chosenArticle: Article)
    }
}
